### PR TITLE
provider/archive: Converting to datasource.

### PR DESCRIPTION
**Pull request by BSick7**
Since the output file is always generated, you will get flapping between runs.

It makes sense to just convert to datasource and always generate a new archive file, but only show as new if sha of output file is different.

### CODE REVIEW

**Comment by kwilczynski** [comment]
@BSick7 nice!

This also takes care about the clean-up issue, I've been talking to @stack72 about.

File: builtin/providers/archive/data_source_archive_file.go [hunk 1]
```diff
@@ -68,20 +63,26 @@ func resourceArchiveFile() *schema.Resource {
...
 
-func resourceArchiveFileCreate(d *schema.ResourceData, meta interface{}) error {
-	if err := resourceArchiveFileUpdate(d, meta); err != nil {
+func dataSourceFileRead(d *schema.ResourceData, meta interface{}) error {
+	outputPath := d.Get("output_path").(string)
+
+	outputDirectory := path.Dir(outputPath)
+	if outputDirectory != "" {
+		if _, err := os.Stat(outputDirectory); err != nil {
+			if err := os.MkdirAll(outputDirectory, 0777); err != nil {
```
Comment (kwilczynski) [inline builtin/providers/archive/data_source_archive_file.go]: I would prefer '0755' to be here.

Comment (BSick7) [inline builtin/providers/archive/data_source_archive_file.go]: Good stuff.
I should probably do 0644 actually.

Comment (kwilczynski) [inline builtin/providers/archive/data_source_archive_file.go]: If you do, you won't be able to enter the directory created. :)

**Comment by BSick7** [comment]
@kwilczynski Done!

**Comment by BSick7** [comment]
@apparentlymart rebased to fix build, ready for review

**Comment by Ninir** [comment]
@BSick7 For informative purposes, could you explain a bit how did you avoid the flapping between runs from the code? (and was `flapping` about the zip sha sum always changing?)
Trying to learn go and get deeper into Terraform... thanks!

**Comment by BSick7** [comment]
@Ninir The original resource code had an Exists function that returned true only if the archive file already existed.  Since the file never survives the current plan in Atlas, it would appear as if it needed to always create the resource.

Since it's now a datasource, the archive file is always generated.  The sha is set as the id at the end of the read.  This is what terraform uses to decipher whether the datasource appears new.

**Comment by Ninir** [comment]
@BSick7 Ok got it, thank you for the explanation! 👍

**Comment by jen20** [comment]
Hi @BSick7! This looks like a nice improvement - I'm going to pull it locally and verify the tests all still pass before merging. Thanks for the pull request!

File: builtin/providers/archive/data_source_archive_file.go [hunk 2]
```diff
@@ -4,20 +4,15 @@ import (
 	"crypto/sha1"
 	"encoding/hex"
 	"fmt"
+	"github.com/hashicorp/terraform/helper/schema"
```
Comment (jen20) [inline builtin/providers/archive/data_source_archive_file.go]: This needs `goimports` running. I'll do this when I pull locally.

**Comment by jen20** [comment]
HI @BSick7! I've pulled this into a branch in our repository (archive-data-source) and fixed up a few minor things. One thing I have noticed is that the tests are failing:

```
=== RUN   TestAccArchiveFile_Basic
--- FAIL: TestAccArchiveFile_Basic (0.02s)
       	testing.go:265: Step 0 error: Check failed: Check 2/2 error: Not found: archive_file.foo
       	testing.go:329: Error destroying resource! WARNING: Dangling resources
       		may exist. The full state and error is shown below.

       		Error: Check failed: Check 1/1 error: found file expected to be deleted: zip_file_acc_test.zip

       		State: <no state>
```

This makes sense since we never delete the file - we should determine whether that is correct behaviour or not and update the destroy check if so.

**Comment by apparentlymart** [comment]
Data sources don't have a `Delete` action since they are assumed to not have side-effects. It feels a bit weird to create something persistent in a data source, honestly... and the fact that it can't get deleted is one example of that weirdness.

Perhaps we could make this feel better as a data source by making the file it creates be explicitly a temporary file (e.g. in `$TEMPDIR`) and clean it up automatically on Terraform exit, rather than putting it in a place that makes it feel persistent and leaving it there. It being temporary feels more natural in a different way, too: local resources are not likely to persist from one Terraform run to another if Terraform isn't always run on the same machine, and so making it a data source and temporary makes it explicit that the file won't persist and needs to be re-generated on each run.

**Comment by BSick7** [comment]
I like the idea of using `$TEMPDIR`.  Does this construct currently exist?

**Comment by apparentlymart** [comment]
It looks like we use `ioutil.TempFile` in some places, which achieves part of the problem. The challenge here would be doing the cleanup step.

When I originally suggested this I was thinking about the explicit provider "Close" step that we ostensibly support, which could in principle allow the archive provider to keep a list of all the files it created and then delete them once it is closed. However, digging deeper into the code I see that `ResourceProviderCloser` seems to be a bit of a mirage: as far as I can tell, no provider actually implements this interface because `helper/schema` itself doesn't seem to implement it.

So this doesn't seem as straightforward as it did when I originally suggested it. We could try to add support for a `CloseFunc` callback to the `ResourceProvider` implementation in `helper/schema`, which would act as the close counterpart to `ConfigureFunc`... though honestly I'm not sure if that happens late enough for what we want here, since Terraform will probably execute the "close provider" action as soon as all of the resources from the provider have been dealt with, and not necessarily wait until all of the dependencies of the resources have been dealt with.

So I think perhaps my suggestion can't be done right now without some core Terraform work. We could potentially just start using `ioutil.TempFile` and accept that these files don't get cleaned by Terraform, presuming that some process outside of Terraform will clean it up... but that's not great if the archive files created will tend to be quite large in size, which I think is the case for bundles for AWS Lambda.

**Comment by BSick7** [comment]
@jen20 @apparentlymart 
ran goimports

I'm still a little confused by the dangling resource issue.
If a datasource doesn't have a `Delete`, why would the test expect to delete?
Should I drop the resource shim?

**Comment by BSick7** [comment]
Now fixes #8501 by adding `output_base64sha256` exported attribute.

**Comment by coen-hyde** [comment]
It'd be nice if this archive data source didn't require writing the file to disk. Instead there could be a source output which could be fed into other resources, eg aws_lambda_function.

**Comment by Ninir** [comment]
Can't wait to have this, so many issues will be resolved with this! thank you @BSick7 :)

**Comment by BSick7** [comment]
@jen20 I believe I have fixed this.
Any chance you could verify my acceptance test again?

**Comment by jen20** [comment]
Thanks @BSick7! One of us will review this over the next day or two for inclusion in the next version of Terraform.

**Comment by stack72** [comment]
Hi @BSick7 

This works now as expected :)

```
% make testacc TEST=./builtin/providers/archive                                                                                                                                                       130 ↵
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/10/25 15:58:29 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/archive -v  -timeout 120m
=== RUN   TestAccArchiveFile_Basic
--- PASS: TestAccArchiveFile_Basic (0.06s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestZipArchiver_Content
--- PASS: TestZipArchiver_Content (0.00s)
=== RUN   TestZipArchiver_File
--- PASS: TestZipArchiver_File (0.00s)
=== RUN   TestZipArchiver_Dir
--- PASS: TestZipArchiver_Dir (0.00s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/archive	0.077s
```

Thanks for the work here!

**Comment by Quentin-M** [comment]
This actually breaks a few usage since there is now no way to have it depend on another resource.

**Pull request merged**